include/dba.php

```diff
@@ -497,6 +441,272 @@ public function any_value_fallback($sql) {
+				if (mysql_errno(self::$dbo->db)) {
+					self::$dbo->error = mysql_error(self::$dbo->db);
+					self::$dbo->errorno = mysql_errno(self::$dbo->db);
+				}
+				break;
+		}
+
+		$a->save_timestamp($stamp1, 'database');
+
+		if (x($a->config,'system') && x($a->config['system'], 'db_log')) {
+
+			$stamp2 = microtime(true);
+			$duration = (float)($stamp2 - $stamp1);
+
+			if (($duration > $a->config["system"]["db_loglimit"])) {
+				$duration = round($duration, 3);
+				$backtrace = debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS);
+				@file_put_contents($a->config["system"]["db_log"], datetime_convert()."\t".$duration."\t".
+						basename($backtrace[1]["file"])."\t".
+						$backtrace[1]["line"]."\t".$backtrace[2]["function"]."\t".
+						substr($sql, 0, 2000)."\n", FILE_APPEND);
+			}
+		}
+		return $retval;
+	}
+
+	/**
+	 * @brief Executes a prepared statement
+	 *
+	 * @param string $sql SQL statement
+	 * @return boolean Was the query successfull?
+	 */
+	static public function e($sql) {
+		$a = get_app();
+
+		$stamp = microtime(true);
+
+		$args = func_get_args();
+
+		$stmt = call_user_func_array('self::p', $args);
+
+		if (is_bool($stmt)) {
+			$retval = $stmt;
+		} elseif (is_object($stmt)) {
+			$retval = true;
+		} else {
+			$retval = false;
+		}
+
+		self::close($stmt);
+
+		$a->save_timestamp($stamp, "database_write");
+
+		return $retval;
+	}
+
+	/**
+	 * @brief Check if data exists
+	 *
+	 * @param string $sql SQL statement
+	 * @return boolean Are there rows for that query?
+	 */
+	static public function exists($sql) {
+		$args = func_get_args();
+
+		$stmt = call_user_func_array('self::p', $args);
+
+		if (is_bool($stmt)) {
+			$retval = $stmt;
+		} else {
+			$retval = (self::rows($stmt) > 0);
+		}
+
+		self::close($stmt);
+
+		return $retval;
+	}
+
+	/**
+	 * @brief Returns the number of rows of a statement
+	 *
+	 * @param object Statement object
+	 * @return int Number of rows
+	 */
+	static public function num_rows($stmt) {
+		switch (self::$dbo->driver) {
+			case 'pdo':
+				return $stmt->rowCount();
+			case 'mysqli':
+				return $stmt->num_rows;
+			case 'mysql':
+				return mysql_num_rows($stmt);
+		}
+		return 0;
+	}
+
+	/**
+	 * @brief Fetch a single row
+	 *
+	 * @param object $stmt statement object
+	 * @return array current row
+	 */
+	static public function fetch($stmt) {
+		if (!is_object($stmt)) {
+			return false;
+		}
+
+		switch (self::$dbo->driver) {
+			case 'pdo':
+				return $stmt->fetch(PDO::FETCH_ASSOC);
+			case 'mysqli':
+				// This code works, but is slow
+
+				// Bind the result to a result array
+				$cols = array();
+
+				$cols_num = array();
+				for ($x = 0; $x < $stmt->field_count; $x++) {
+					$cols[] = &$cols_num[$x];
+				}
+
+				call_user_func_array(array($stmt, 'bind_result'), $cols);
+
+				if (!$stmt->fetch()) {
+					return false;
+				}
+
+				// The slow part:
+				// We need to get the field names for the array keys
+				// It seems that there is no better way to do this.
```

Comment:
Have you considered [`msqli_result::fetch_assoc()`](http://php.net/manual/en/mysqli.quickstart.prepared-statements.php#example-1808)?